src/codecatalyst/devEnv.ts

```diff
@@ -0,0 +1,206 @@
+
+/** If we should be sending the dev env heartbeats to track user activity */
+export function shouldTrackUserActivity(maxInactivityMinutes: DevEnvironment['inactivityTimeoutMinutes']): boolean {
+    // This value is a static value initialized when the dev env is first created, if it is updated the dev env is restarted
+    // For more info see: https://docs.aws.amazon.com/codecatalyst/latest/APIReference/API_UpdateDevEnvironment.html#codecatalyst-UpdateDevEnvironment-request-inactivityTimeoutMinutes
+    return maxInactivityMinutes > 0
+}
+
+export class InactivityMessages {
+    static #instance: InactivityMessages | undefined
+
+    static get instance(): InactivityMessages {
+        return (this.#instance ??= new InactivityMessages())
+    }
+
+    /**
+     * Sets up all messages that are displayed to the user when the dev env is inactive and starts to get close to shutdown.
+     *
+     * @param inactivityMinutes The # of minutes a CodeCatalyst Dev Env can be inactive till it shuts down.
+     * @param devEnv The DevEnv client
+     * @returns
+     */
+    async setupMessages(
+        inactivityMinutes: DevEnvironment['inactivityTimeoutMinutes'],
+        devEnv: DevEnvClient,
+        warningMessage: InactivityWarningMessage,
+        shutdownMessage: InactivityShutdownMessage
+    ) {
+        const maxInactivityMinutes = inactivityMinutes
+
+        devEnv.onUserActivity(() => {
+            const userIsActive = () => devEnv.updateUserActivity()
+            warningMessage.setupMessage(maxInactivityMinutes, userIsActive)
+            shutdownMessage.setupMessage(maxInactivityMinutes, userIsActive, () =>
+                warningMessage.clearExistingMessage()
+            )
+        })
+
+        devEnv.updateUserActivity()
+    }
+}
+
+export class InactivityWarningMessage {
+    private beforeWarningMessageShownTimeout: NodeJS.Timeout | undefined
+    private currentWarningMessageTimeout: Timeout | undefined
+
+    static #instance: InactivityWarningMessage | undefined
+
+    static get instance(): InactivityWarningMessage {
+        return (this.#instance ??= new InactivityWarningMessage())
+    }
+
+    /**
+     * @WARNING Only use this for testing purposes, use {@link instance} otherwise.
+     *
+     * @param relativeMinuteMillis What a minute is in the context of testing
+     */
+    constructor(private readonly relativeMinuteMillis?: number) {}
+
+    async setupMessage(maxInactivityMinutes: number, userIsActive: () => Promise<any> | any): Promise<void> {
+        this.clearExistingMessage()
+
+        const showMessageAfterMinutes = this.getMinutesUntilShowWarningMessage(maxInactivityMinutes)
+
+        this.beforeWarningMessageShownTimeout = globals.clock.setTimeout(() => {
+            const inactiveMinutes = showMessageAfterMinutes
+            const remainingMinutes = maxInactivityMinutes - inactiveMinutes
+            this.showInactivityWarningMessage(inactiveMinutes, remainingMinutes, userIsActive)
```

Comment:
```suggestion
            const remainingMinutes = maxInactivityMinutes - showMessageAfterMinutes
            this.showInactivityWarningMessage(showMessageAfterMinutes, remainingMinutes, userIsActive)
```

Or rename `showMessageAfterMinutes`